### PR TITLE
MNT-20764: Searches fail for users who are members of groups where the authorityName contains double quotes

### PR DESCRIPTION
**Pull request by amohammedalfresco**
- Throw exceptions when a group name contain invalid character (`'/', '\\', '\n', '\r', '"'`)

### CODE REVIEW

**Comment by amohammedalfresco** [comment]
Closed #919 due to EOL issues

File: src/test/java/org/alfresco/repo/security/authority/AuthorityServiceTest.java [hunk 1]
```diff
@@ -252,6 +252,20 @@ public void testMNT10533() throws Exception
         }
     }
 
+    public void testCreateAuthorityWithIllegalCharacters() throws Exception
+    {
+        char[] illegalCharacters = {'/', '\\', '\n', '\r', '"'};
+        for (char illegalCharacter : illegalCharacters)
+        {
+            String GROUP_NAME = "testGroupNameWith" + illegalCharacter;
+            try {
```
Comment (killerboot) [inline src/test/java/org/alfresco/repo/security/authority/AuthorityServiceTest.java:261]: Please check your code style. Braces are in wrong places.

File: src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java [hunk 2]
```diff
@@ -1046,6 +1046,17 @@ else if (type.isPrefixed())
             final String authorityName = prefix + shortName;
             if (shortName.startsWith(prefix))
             {
+                if (type == AuthorityType.GROUP)
```
Comment (killerboot) [inline src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java:1049]: Are these characters allowed for users?

File: src/test/java/org/alfresco/repo/security/authority/AuthorityServiceTest.java [hunk 3]
```diff
@@ -351,9 +365,14 @@ public void testZones()
     @Category(RedundantTests.class)
     public void test_ETWOTWO_400()
```
Comment (killerboot) [inline src/test/java/org/alfresco/repo/security/authority/AuthorityServiceTest.java:366]: I think this test needs to be deleted as it is essentially the same as testCreateAuthorityWithIllegalCharacters now.

File: src/test/java/org/alfresco/repo/security/authority/AuthorityServiceTest.java (hunk 1 above)
Comment (amohammedalfresco) [reply src/test/java/org/alfresco/repo/security/authority/AuthorityServiceTest.java]: Thanks for the spotting it, just linted it.

File: src/test/java/org/alfresco/repo/security/authority/AuthorityServiceTest.java (hunk 3 above)
Comment (amohammedalfresco) [reply src/test/java/org/alfresco/repo/security/authority/AuthorityServiceTest.java]: Removed the test.

File: src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java (hunk 2 above)
Comment (amohammedalfresco) [reply src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java:1049]: I thought of initially adding it for the `AuthorityType.USER` but the [createAuthority](https://github.com/Alfresco/alfresco-repository/blob/master/src/main/java/org/alfresco/repo/security/authority/AuthorityServiceImpl.java#L645) method calls another method [checkTypeIsMutable](https://github.com/Alfresco/alfresco-repository/blob/master/src/main/java/org/alfresco/repo/security/authority/AuthorityServiceImpl.java#L527) and it has explicit check for either `AuthorityType.GROUP` or `AuthorityType.ROLE` which makes me to believe this `createAuthority` method may not be creating `AuthorityType.USER`.  Hence, I only guarded the `checkTypeIsMutable` method with `AuthorityType.Group`.

Comment (killerboot) [reply src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java:1049]: I think it is required to extend the scope of this change to make sure that Users and Groups can not be created with wrong characters using Java APIs.

File: src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java [hunk 4]
```diff
@@ -1656,6 +1657,28 @@ public void onUpdateProperties(NodeRef nodeRef, Map<QName, Serializable> before,
         }
     }
 
+    public void onCreateNode(ChildAssociationRef childAssocRef)
+    {
+        // Restrict creation of group name that contain invalid characters.
+        NodeRef groupRef = childAssocRef.getChildRef();
```
Comment (killerboot) [inline src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java:1663]: Please change the variables to `authorityRef` and `authorityName`

File: src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java [hunk 5]
```diff
@@ -1656,6 +1657,28 @@ public void onUpdateProperties(NodeRef nodeRef, Map<QName, Serializable> before,
...
     }
 
+    public void onCreateNode(ChildAssociationRef childAssocRef)
+    {
+        // Restrict creation of group name that contain invalid characters.
+        NodeRef groupRef = childAssocRef.getChildRef();
+
+        String groupName = (String) this.nodeService.getProperty(groupRef, ContentModel.PROP_AUTHORITY_NAME);
+
+        if (getAuthorityContainer().equals(childAssocRef.getParentRef()))
```
Comment (killerboot) [inline src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java:1667]: Is this check required?

File: src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java (hunk 4 above)
Comment (amohammedalfresco) [reply src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java]: Done.

File: src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java (hunk 5 above)
Comment (amohammedalfresco) [reply src/main/java/org/alfresco/repo/security/authority/AuthorityDAOImpl.java]: Removed the not needed check.

**Pull request merged**